plugin.py

```diff
@@ -0,0 +1,136 @@
+import sublime
+
+import os
+import re
+import base64
+from urllib.parse import urlparse, parse_qs
+import weakref
+
+from LSP.plugin.core.protocol import Request, Location
+from LSP.plugin.core.typing import Any, Callable, Dict, Iterable, List, Mapping, Optional, TypedDict, Union
+from LSP.plugin.core.registry import LspTextCommand
+from LSP.plugin import register_plugin, unregister_plugin
+from LSP.plugin import AbstractPlugin, ClientConfig, SessionBufferProtocol
+from LSP.plugin import WorkspaceFolder
+from LSP.plugin.core.types import Optional, Any, Tuple, List
+from lsp_utils import ApiWrapperInterface, NpmClientHandler
+
+class LspGrammarlyCommand(LspTextCommand):
+    session_name = 'grammarly'
+
+class LspGrammarlyExecuteLoginCommand(LspGrammarlyCommand):
+    cmd = "$/getOAuthUrl"
+    cmd_callback = "$/handleOAuthCallbackUri"
+
+    internal_redirect_uri = "vscode://znck.grammarly/auth/callback"
+    external_redirect_uri = internal_redirect_uri
+    redirect_uri = internal_redirect_uri
+
+    def run(self, edit: sublime.Edit) -> None:
+        def run_async() -> None:
+            session = self.session_by_name()
+            if session:
+                request = Request(self.cmd, self.redirect_uri, None, progress=True)
+                self.weaksession = weakref.ref(session)
+                session.send_request(request, self._handle_response,
+                    lambda r: print(self.cmd, r))
+            else:
+                self._handle_resolve_response_async(None, item)
+        sublime.set_timeout_async(run_async)
+
+    def _handle_response(self, response: Union[List[Location], None]) -> None:
+        link = re.sub("state=[^&]*", "state=" + base64.b64encode(bytes(self.external_redirect_uri, 'utf-8')).decode('ascii'), str(response))
+        session = self.weaksession()
+        if session:
+            session.window.show_input_panel("Open this url in the browser and paste in its place the redirection:", link,
+                                     lambda r: self._send_response(r), None, lambda: self._send_response(None))
+        
+    def _send_response(self, input: Optional[str]) -> None:
+        if input:
+            res = urlparse(input)
+            assert(res.scheme == 'vscode')
+            request = Request(self.cmd_callback, input, None, progress=True)
+            session = self.weaksession()
+            if session:
+                session.send_request(request, lambda p: print("Connected." + str(p)),
+                    lambda r: print("Invalid authorization-response url"))
```

Comment:
Can we use `sublime.message_dialog` & `sublime.error_message` ([API](http://www.sublimetext.com/docs/api_reference.html#sublime)) for such errors. It's not very user friendly to just print to the console as user is not likely to look there.